File: python/algo/202509/algo2601.py
import bisect

N = 1003
is_prime = [True] * N
is_prime[0] = is_prime[1] = False
primes = []

for i in range(2, N):
    if is_prime[i]:
        primes.append(i)
        # 标记i的倍数为非质数
        for j in range(i * i, N, i):
            is_prime[j] = False


# 数学算法 - 数论 - 1.2 预处理质数（筛质数）
class Solution:
    def primeSubOperation(self, nums):
        n = len(nums)
        # 从后往前处理数组
        for i in range(n-2, -1, -1):
            if nums[i] < nums[i+1]:
                continue
            # 计算需要减小的差值
            diff = nums[i] - nums[i+1]
            # 找到第一个大于diff的质数
            idx = bisect.bisect_right(primes, diff)
            if (idx == len(primes)) :return False
            p = primes[idx]
            nums[i] -= p
            if nums[i] <= 0:
                return False
        return True

File: cplus/leetcode/202509/algo2601.cpp
#include <bits/stdc++.h>
using namespace std;

const int N = 1e3 + 1;
bool is_prime[N];
vector<int> primes;
int init = [] {
    memset(is_prime, true, sizeof(is_prime));
    is_prime[0] = false; is_prime[1] = false;
    for (int i = 2; i < N; ++i) {       // 需要统计质数数量，此处不能筛至平方根
        if (is_prime[i]) {
            primes.push_back(i);
            for (long long j = 1LL * i * i; j < N; j += i) {
                is_prime[j] = false;
            }
        }
    }
    return 0;
}();

// 数学算法 - 数论 - 1.2 预处理质数（筛质数）
class Solution {
public:
    bool primeSubOperation(vector<int>& nums) {
        int n = nums.size();
        for (int i = n-2; i>=0; i--) {
            if (nums[i] < nums[i+1]) continue;
            else {
                auto diff = nums[i] - nums[i+1];
                auto it = ranges::upper_bound(primes, diff);
                if (it != primes.end()) nums[i] -= *it;
                else return false;
                
                if (nums[i] <= 0) return false;
            }
        }

        return true;
    }
};

